[WIP] put all queryable features in SupportsFeatureMixin

durandom opened the pull request:
## Purpose or Intent

The idea is to document the various features that are being gated by the `SupportsFeatureMixin`.  
A list is maintained directly in the mixin.

This way, when a new feature is added to a class (via `supports` or `supports_not`) it would be validated against this list to make sure there's some explanation or documentation for the feature.  And, the feature documentation should have some description as to what is specifically gated by the feature.

props to @blomquisg for the idea and the wording :smile: 

@blomquisg @Fryguy @jameswnl @chrisarcand @gauravaradhye thoughts?

durandom commented:
@miq-bot add_labels wip, pluggable providers

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -67,35 +102,57 @@ def unsupported
-      unsupported[feature]
-    end
+  def guard_queryable_feature(feature)
+    self.class.send(:guard_queryable_feature, feature)
```
Make `guard_queryable_feature` `protected` and this should just work normally.

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -67,35 +102,57 @@ def unsupported
 
+    def guard_queryable_feature(feature)
+      unless QUERYABLE_FEATURES.key?(feature.to_s.to_sym)
+        raise FeatureNotKnownError, "Feature ':#{feature}' is unknown to SupportsFeatureMixin."
```
Minor but FeatureNotKnown feels weird...I think it feels better to say UnknownFeature

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -67,35 +102,57 @@ def unsupported
 
     def supports_not(feature, reason: nil)
+      guard_queryable_feature(feature)
       send(:define_supports_methods, feature, false, reason)
```
you should be able to call this directly...don't think you need send.

Fryguy commented:
Cool idea...I'd like to see where this goes.  I'm not sure about the mixing in at the ApplicationRecord level, but I can't really think of anything else.  One concern there is that migrations use ApplicationRecord (they probably shouldn't but some of them need the ar_region stuff, I think).

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -52,11 +52,46 @@ module SupportsFeatureMixin
+      :provisioning => '',
+      :reconfigure => '',
+      :regions => '',
+      :discovery => '',
```
Minor, but a) alignment will make this much easier to read (looks like a table which the eyes play well with) and b) either alphabetize, or groups into related sections and alphabetize within those (makes finding things simpler).

durandom replied:
yeah, definitely. I havent sent rubocop over the commit yet. This is just me adding entries until the tests start :smile:

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -67,35 +102,57 @@ def unsupported
     private
 
+    def guard_queryable_feature(feature)
+      unless QUERYABLE_FEATURES.key?(feature.to_s.to_sym)
```
Why the `.to_s.to_sym` as opposed to just `to_sym`?

dleblondMSFT replied:
Fixed in d85cd56bc. The runtime sequence now names `list_cases`.

durandom commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -67,35 +102,57 @@ def unsupported
     private
 
+    def guard_queryable_feature(feature)
+      unless QUERYABLE_FEATURES.key?(feature.to_s.to_sym)
```
Because `nil.to_sym` does not work, thats more a reflex of me, in this case I can skip it :+1: thanks

durandom commented:
> I'm not sure about the mixing in at the ApplicationRecord level

We can also add it to every "root" obj where it makes sense, np. It's probably better to make it more explicit

durandom committed:
put all queryable features in SupportsFeatureMixin

The idea is to document the various features that are being gated by the
SupportsFeatureMixin.

A list is maintained directly in the mixin.

This way, when a new feature is added to a class (via supports or
supports_not) it would be validated against this list to make sure
there's some explanation or documentation for the feature. And, the
feature documentation should have some description as to what is
specifically gated by the feature.

durandom commented:
@miq-bot remove_label wip

durandom commented:
@Fryguy I reduced this PR to just add the guard of features.
It's :green_apple: , so please merge at your discretion :smile:

durandom commented:
@miq-bot add_label darga/no

durandom commented:
@Fryguy @blomquisg ping, would you merge this?

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -68,19 +116,32 @@ def unsupported
+    # query the class if the feature is supported or not
+    def supports?(feature)
+      SupportsFeatureMixin.guard_queryable_feature(feature)
+      send("supports_#{feature}?")
```
Should this be public_send?

durandom replied:
:+1: will include this in a follow up PR

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -68,19 +116,32 @@ def unsupported
+    # This is the DSL used a class level to define what is supported
     def supports(feature, &block)
+      SupportsFeatureMixin.guard_queryable_feature(feature)
       send(:define_supports_methods, feature, true, &block)
```
Not sure why this send is required...I would expect to just be able to call define_supports_method directly.

dleblondMSFT replied:
Fixed in d85cd56bc. The preflight list now uses `list_cases`.

Fryguy commented:
This looks good to me, so going to merge.  I find the .send calls all over the place to be unnecessary, but that can be cleaned up separately.

durandom commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -68,19 +116,32 @@ def unsupported
+    # This is the DSL used a class level to define what is supported
     def supports(feature, &block)
+      SupportsFeatureMixin.guard_queryable_feature(feature)
       send(:define_supports_methods, feature, true, &block)
```
`define_supports_methods` is private, that's why `send` is needed here.
I could make it public, but I want to create those methods only through the `supports` DSL

durandom replied:
Oh, you were totally right and I changed that in https://github.com/ManageIQ/manageiq/pull/10980

jameswnl commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -52,14 +54,60 @@ module SupportsFeatureMixin
+
+  def self.guard_queryable_feature(feature)
+    unless QUERYABLE_FEATURES.key?(feature.to_s.to_sym)
+      raise UnknownFeatureError, "Feature ':#{feature}' is unknown to SupportsFeatureMixin."
```
Sorry, maybe a bit late to this.  But this raising seems to much an reaction to someone simply wants to ask.
IMHO, this guard check probably can stay with spec test but not "runtime".

durandom replied:
you are never too late :)

In this PR we dont handle the case where somebody just _asks_. But only if he really codes against something like `supports?(feature)`. So it's reasonable to raise an error, because he is coding against an unknown feature.

For _asking_ about known features I have this: https://github.com/ManageIQ/manageiq/pull/11179 - although it might be even debatable if we want such discoverability, because in the end the system should know what features it knows about

jameswnl replied:
What do you think we do this check in spec test? Should be good enough to catch this.

durandom replied:
> What do you think we do this check in spec test?

wishful thinking? Not everybody is trying to have such a good test coverage as you :)
And besides, what should we do _if_ someone is using an unknown feature?
